From e31c149c27e5745be964577ed8f2f1fcbfa62bf4 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Thu, 7 Aug 2025 22:45:33 -0400
Subject: [PATCH] ntdll: Implement RtlAreLongPathsEnabled.

(cherry picked from commit 1cb81299395ff9375db2b9b6ca33c86cce626d10)
---
 dlls/ntdll/ntdll.spec | 1 +
 dlls/ntdll/rtl.c      | 8 ++++++++
 dlls/ntdll/unix/env.c | 2 ++
 3 files changed, 11 insertions(+)

diff --git a/dlls/ntdll/ntdll.spec b/dlls/ntdll/ntdll.spec
index 32b1d5e3d6c..e08174d90b2 100644
--- a/dlls/ntdll/ntdll.spec
+++ b/dlls/ntdll/ntdll.spec
@@ -524,6 +524,7 @@
 @ stdcall RtlAreAnyAccessesGranted(long long)
 @ stdcall RtlAreBitsClear(ptr long long)
 @ stdcall RtlAreBitsSet(ptr long long)
+@ stdcall RtlAreLongPathsEnabled()
 # @ stub RtlAssert2
 @ stdcall RtlAssert(ptr ptr long str)
 @ stdcall RtlBarrier(ptr long)
diff --git a/dlls/ntdll/rtl.c b/dlls/ntdll/rtl.c
index a2b438b1418..0e2e6a91b51 100644
--- a/dlls/ntdll/rtl.c
+++ b/dlls/ntdll/rtl.c
@@ -1655,6 +1655,14 @@ void WINAPI RtlGetDeviceFamilyInfoEnum(ULONGLONG *version, DWORD *family, DWORD
     if (form) *form = DEVICEFAMILYDEVICEFORM_UNKNOWN;
 }
 
+/*********************************************************************
+ *           RtlAreLongPathsEnabled [NTDLL.@]
+ */
+BOOL WINAPI RtlAreLongPathsEnabled(void)
+{
+    return NtCurrentTeb()->Peb->IsLongPathAwareProcess;
+}
+
 /*********************************************************************
  *           RtlConvertDeviceFamilyInfoToString [NTDLL.@]
  */
diff --git a/dlls/ntdll/unix/env.c b/dlls/ntdll/unix/env.c
index 3145868e4ba..104b1eb90f8 100644
--- a/dlls/ntdll/unix/env.c
+++ b/dlls/ntdll/unix/env.c
@@ -1874,6 +1874,7 @@ static void init_peb( RTL_USER_PROCESS_PARAMETERS *params, void *module )
     peb->ImageSubSystem             = main_image_info.SubSystemType;
     peb->ImageSubSystemMajorVersion = main_image_info.MajorSubsystemVersion;
     peb->ImageSubSystemMinorVersion = main_image_info.MinorSubsystemVersion;
+    peb->IsLongPathAwareProcess     = TRUE;
 
 #ifdef _WIN64
     if (!is_machine_64bit( main_image_info.Machine ))
@@ -1909,6 +1910,7 @@ static void init_peb( RTL_USER_PROCESS_PARAMETERS *params, void *module )
         wow_peb->ImageSubSystemMajorVersion      = peb->ImageSubSystemMajorVersion;
         wow_peb->ImageSubSystemMinorVersion      = peb->ImageSubSystemMinorVersion;
         wow_peb->SessionId                       = peb->SessionId;
+        wow_peb->IsLongPathAwareProcess          = peb->IsLongPathAwareProcess;
     }
 }
 
